File: 2018_KAKAO_BLIND_RECRUITMENT/프렌즈4블록.js
function solution(m, n, board) {
    var answer = 0;
    // console.log(board)
    for (let i = 0; i < board.length; i++) {
        const temp = board[i];
        board[i] = temp.split('')
    }
    // console.log(board)

    // let flag = false
    while (true) {
        let flag = true
        const will_delete = [];
        for (let i = 0; i < m - 1; i++) {
            for (let j = 0; j < n - 1; j++) {
                const a = board[i][j]
                if (board[i + 1][j] === a && board[i][j+1] === a && board[i+1][j+1] === a && a !== 0) {
                    flag = false
                    // console.log('여기오냐 계속?')
                    // 똑같은 실수를...
                    will_delete.push([i, j])
                }
            }
        }
        if (flag === true) {
            break
        }

        // console.log(will_delete)
        will_delete.forEach(dell => {
            const y = dell[0]
            const x = dell[1]
            for (let i = 0; i < 2; i++) {
                for (let j = 0; j < 2; j++) {
                    const idy = y + i
                    const jdx = x + j
                    if (board[idy][jdx] !== 0) {
                        answer += 1
                    }
                    board[idy][jdx] = 0
                }
            }
        });
        // console.log(board)

        // 이거는 좀 대박인거같은데... 바꿔도 주소값 차이 없겠지??...
        // const new_board = Array(m).fill(Array(n).fill(0))
        const new_board = new Array(m);
        for (let q = 0; q < m; q++) {
            new_board[q] = new Array(n).fill(0);
        }
        // 이렇게 하면 된다!

        for (let i = 0; i < n; i++) {
            const stack = Array()
            for (let j = m - 1; j >= 0; j--) {
                if (board[j][i] !== 0) {
                    stack.push(board[j][i])
                }
            }
            // console.log(stack)
            for (let k = 0; k < stack.length; k++) {
                new_board[m - 1 - k][i] = stack[k]
            }
        }
        // console.log(new_board)
        // wow 드디어...ㄷㄷ
        board = new_board
        // console.log(board)
        }
    console.log(answer)
    return answer;
}

solution(4, 5, ['CCBDE', 'AAADE', 'AAABF', 'CCBBF'])
// solution(6, 6, ['TTTANT', 'RRFACC', 'RRRFCC', 'TRRRAA', 'TTMMMF', 'TMMTTJ'])

File: 2018_KAKAO_BLIND_RECRUITMENT/프렌즈4블록.py
def solution(m, n, board):
    answer = 0
    board_map = [0] * m
    for i in range(m):
        board_map[i] = list(board[i])
    # print(board_map)

    stack = [0]
    while stack != []:
        stack = []
        for i in range(m - 1):
            for j in range(n - 1):
                if board_map[i][j] == board_map[i + 1][j] == board_map[i][j + 1] == board_map[i + 1][j + 1] != 0: # 0은 아니어야 stack에 채우는 것을 멈춘당!
                    stack.append((i, j))
        # stack에 점 찍어둔 것 처리하기. 없어지는 갯수를 세어야해서 중복 주의
        for k in range(len(stack)):
            i, j = stack[k]
            for ii in range(2):
                for jj in range(2):
                    if board_map[i + ii][j + jj] != 0:
                        answer += 1
                    board_map[i + ii][j + jj] = 0

        # print(board_map)
        # 뿌요뿌요처럼 내리는 로직
        board_map2 = [ [ 0 ] * n for _ in range(m) ]
        for j in range(n):
            y = m - 1
            for i in range(m - 1, -1, -1):
                if board_map[i][j] != 0:
                    board_map2[y][j] = board_map[i][j]
                    y -= 1
        board_map = board_map2
        # 저번에 짠거랑은 좀 다른데 이것도 되겠지?
        # print(stack)

    # print(answer)
    return answer


solution(6, 6, ["TTTANT", "RRFACC", "RRRFCC", "TRRRAA", "TTMMMF", "TMMTTJ"])